contracts/modules/Experimental/STO/POLYPeggedSTO.sol

```diff
@@ -0,0 +1,572 @@
+    }
+
+    /**
+     * @dev Modifies addresses used as wallet and treasury wallet
+     * @param _wallet Address of wallet where funds are sent
+     */
+    function modifyWalletAddress(address payable _wallet) external {
+        //can be modified even when STO started
+        _onlySecurityTokenOwner();
+        _modifyWalletAddress(_wallet);
+    }
+
+    /**
+     * @notice Use to change the treasury wallet
+     * @param _treasuryWallet Ethereum account address to receive unsold tokens
+     * @notice Set to address zero to use dataStore treasuryWallet
+     */
+    function modifyTreasuryWallet(address _treasuryWallet) external {
+        //can be modified even when STO started
+        _onlySecurityTokenOwner();
+        _modifyTreasuryWallet(_treasuryWallet);
+    }
+
+    function _modifyLimits(
+        uint256 _minimumInvestment,
+        uint256 _nonAccreditedLimit,
+        uint256 _maxNonAccreditedInvestors
+    ) internal {
+        minimumInvestment = _minimumInvestment;
+        nonAccreditedLimit = _nonAccreditedLimit;
+        maxNonAccreditedInvestors = _maxNonAccreditedInvestors;
+        emit SetLimits(minimumInvestment, nonAccreditedLimit, maxNonAccreditedInvestors);
+    }
+
+    function _modifyRate(uint256 _rate) internal {
+        require(_rate > 0, "Rate not > 0");
+        rate = _rate;
+        emit SetRate (rate);
+    }
+
+    function _modifyCap(uint256 _cap) internal {
+        require(_cap > 0, "Cap not > 0");
+        cap = _cap;
+        emit SetCap (cap);
+    }
+
+    function _modifyTimes(uint256 _startTime, uint256 _endTime) internal {
+        /*solium-disable-next-line security/no-block-members*/
+        require(_startTime >= now && _endTime > _startTime, "Invalid times");
+        startTime = _startTime;
+        endTime = _endTime;
+        emit SetTimes(_startTime, _endTime);
+    }
+
+    function _modifyWalletAddress(address payable _wallet) internal {
+        require(_wallet != address(0), "Invalid wallet");
+        emit SetWallet(wallet, _wallet);
+        wallet = _wallet;
+    }
+
+    function _modifyTreasuryWallet(address _treasuryWallet) internal {
+        emit SetTreasuryWallet(treasuryWallet, _treasuryWallet);
+        treasuryWallet = _treasuryWallet;
+    }
+
+    ////////////////////
+    // STO Management //
+    ////////////////////
+
+    /**
+     * @notice Finalizes the STO and mints remaining tokens to treasury address
+     * @notice Treasury address must be whitelisted to successfully finalize
+     * @param _mintUnsoldTokens unsold tokens will be minted to the Treasury wallet if ture
+     */
+    function finalize(bool _mintUnsoldTokens) external {
+        _onlySecurityTokenOwner();
+        require(!isFinalized, "Already finalized");
+        isFinalized = true;
+        if ((_mintUnsoldTokens) && (totalTokensSold < cap)) {
+            address treasury = (treasuryWallet == address(0) ? IDataStore(getDataStore()).getAddress(TREASURY) : treasuryWallet);
```

Comment:
Add one more check for treasury wallet to validate `treasury != 0x0`. It will act as a one more safety measure if something bad happens with dataStore storage.